### PR TITLE
232 web3button

### PR DESCRIPTION
**Pull request by L03TJ3**
# Description

A web3action button used to perform any web3 call.
Can be used with an existing web3 connection infrastructure
or handles default useCases for browser wallet extensions.

- [x] should request to connect if not connected
- [x] should request to switch chain when not connected
- [x] should react to any manual changes as long as isPressed is true
- [x] execute action when above conditions are met

About # (link your issue here)
https://github.com/GoodDollar/GoodProtocolUI/issues/232

# Current issues
- [ ] timeout needs to be cleared, and re-initialized on multiple presses
- [x] fetches read only providers address as account, breaks the flow
- [ ] should handle a provided switch function (opposed/next to generic switchNetwork from useEthers)

# How Has This Been Tested?

Locally in story / ProtocolUI test implementation (Web)

# Checklist:
- [ ] PR title matches follow: (Feature|Bug|Chore) Task Name
- [ ] My code follows the style guidelines of this project
- [ ] I have followed all the instructions described in the initial task (check Definitions of Done)
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I have added reference to a related issue in the repository
- [ ] I have added a detailed description of the changes proposed in the pull request. I am as descriptive as possible, assisting reviewers as much as possible.
- [ ] I have added screenshots related to my pull request (for frontend tasks)
- [ ] I have pasted a gif showing the feature.
- [ ] @mentions of the person or team responsible for reviewing proposed changes

### CODE REVIEW

File: packages/good-design/.storybook/main.js [hunk 1]
```diff
@@ -0,0 +1,66 @@
+const webpack = require('webpack');
```
Comment (sirpy) [inline packages/good-design/.storybook/main.js:1]: i dont think .storybook is to be committed

File: packages/good-design/babel.config.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+module.exports = {
+  "sourceType": "unambiguous",
+  "presets": [
+    "module:metro-react-native-babel-preset",
+    "@babel/preset-flow",
```
Comment (sirpy) [inline packages/good-design/babel.config.js:5]: we dont need flow

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+          return false;
+        }
+      });
+    },
+    [switchNetwork, switchChain]
+  );
+
+  const reset = () => {
+    setLoading(false);
+    setActionText("");
```
Comment (sirpy) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:111]: shouldnt that be set to `text`?

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+  };
+
+  const continueSteps = useCallback(async () => {
+    let cancelled = false;
+    let stepConfirmed: boolean = false;
+
+    // if there is no account always connect
+    // if there is an account but its not a web3provider it mostly indicates a read-only provider
```
Comment (sirpy) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:119]: why would we get a readonly provider from usedapp? explain the !isWeb3 case

File: packages/good-design/src/advanced/web3action/Web3Action.tsx [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+      case "action":
+        try {
+          await web3Action();
+        } finally {
+          reset();
+        }
+        break;
+    }
+
+    if (!stepConfirmed || cancelled) {
```
Comment (sirpy) [inline packages/good-design/src/advanced/web3action/Web3Action.tsx:140]: canceled is never set
case of 'action' stepConfirmed will be false so reset() will be called twice

File: packages/good-design/src/core/buttons/ActionButton.tsx [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+
+interface IBasicButtonProps extends IButtonProps {
+	text: string
+}
+
+export const ButtonAction = ({ text, ...props }: IBasicButtonProps) => {
+	return <Button { ...props } style={styles.buttonAction}>{text}</Button>
+}
+
+const styles = StyleSheet.create({
```
Comment (sirpy) [inline packages/good-design/src/core/buttons/ActionButton.tsx:14]: why aren't we using nativebase styles?

File: packages/good-design/src/core/buttons/ClaimButton.tsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
+import React, { useEffect, useState, useCallback, useMemo } from "react";
```
Comment (sirpy) [inline packages/good-design/src/core/buttons/ClaimButton.tsx:1]: why is the claimbutton in this PR?

File: packages/good-design/src/core/buttons/ClaimButton.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+type FVModalProps = IModalProps & FVFlowProps;
+
+function FVModal({ firstName, method, onClose = noop, ...props }: FVModalProps) {
+  const fvlink = useFVLink();
+  const [loading, setLoading] = useState(false)
+
+  const verify = useCallback(async () => {
+    setLoading(true)
+    await fvlink?.getLoginSig();
```
Comment (sirpy) [inline packages/good-design/src/core/buttons/ClaimButton.tsx:38]: no longer required

File: packages/good-design/src/core/buttons/ClaimButton.tsx [hunk 9]
```diff
@@ -0,0 +1,162 @@
...
+    await fvlink?.getFvSig();
+    setLoading(false)
+
+    switch (method) {
+      case "redirect": {
+        const link = fvlink?.getLink(firstName, document.location.href, false);
```
Comment (sirpy) [inline packages/good-design/src/core/buttons/ClaimButton.tsx:44]: dont use web only code, code must be compatible with react-native

File: packages/good-design/src/core/buttons/ClaimButton.tsx [hunk 10]
```diff
@@ -0,0 +1,162 @@
...
+  if  (loading) return <Spinner/>
+
+  return (
+    <Modal {...props} animationPreset="slide" onClose={onClose}>
+      <View style={styles.containeralt}>
+        <TouchableOpacity style={styles.close} onPress={onClose}>
+          {cross}
+        </TouchableOpacity>
+        <View>
+          <Text color={colors.text1}>To verify your identity you need to sign TWICE with your wallet.</Text>
```
Comment (sirpy) [inline packages/good-design/src/core/buttons/ClaimButton.tsx:75]: only one signature is required

File: packages/good-design/src/core/buttons/ClaimButton.tsx [hunk 11]
```diff
@@ -0,0 +1,162 @@
...
+      <View flex={1} alignItems="center" justifyContent="center">
+        <FVModal method={method} isOpen={showModal} onClose={handleClose} firstName={firstName} />
+      </View>
+      <ButtonAction text={buttonTitle} onPress={handleClaim} />
+    </View>
+  );
+}
+
+const styles = StyleSheet.create({
+  containeralt: {
```
Comment (sirpy) [inline packages/good-design/src/core/buttons/ClaimButton.tsx:133]: use nativebase styles

File: packages/good-design/src/core/layout/Layout.tsx [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+
+interface ILayoutProps {
+	children: ReactNode
+}
+
+export const Layout = ({ children }: ILayoutProps) => <View style={styles.layout}>{children}</View>
+
+export const Title = ({ children }: ILayoutProps) => <Text style={styles.title} color={'#42454A'}>{children}</Text>
+
+const styles = StyleSheet.create({
```
Comment (sirpy) [inline packages/good-design/src/core/layout/Layout.tsx:13]: use nativebase styles

File: packages/good-design/src/stories/Web3Action.stories.tsx [hunk 13]
```diff
@@ -0,0 +1,93 @@
...
+  const [claimText, setClaimText] = useState<string>("Claim UBI");
+
+  const handleClaim = useCallback(async () => {
+    // console.log("HC isWhitelisted -->", { isWhitelisted });
+    // console.log("HC -- account / library", { account, library });
+    if (isWhitelisted) {
+      console.log("isWhitelisted");
+      //todo-fix: this tries to send call from readonly provider, where library already should be Web3Provider
+      // see note above in W3 Wrapper
+      await claimCall.send();
```
Comment (sirpy) [inline packages/good-design/src/stories/Web3Action.stories.tsx:58]: stories should not do real transactions. the web3action button should be jsut demonstrated that it
1. triggers wallet connect
2. ask for chain switch
3. does some TEST action. like showing a modal.

File: packages/good-design/tsconfig.json [hunk 14]
```diff
@@ -0,0 +1,29 @@
+{
+  "compilerOptions": {
+    // Default
+    "target": "es2019",
+    "lib": ["es2019", "dom"],
+    "module": "commonjs",
+    "esModuleInterop": true,
+    "forceConsistentCasingInFileNames": true,
+    "strict": true,
```
Comment (sirpy) [inline packages/good-design/tsconfig.json:9]: i dont think we use strict
make sure there are no typescript warnings when compiling

File: packages/sdk-v2/src/hooks/useEthers.tsx [hunk 15]
```diff
@@ -0,0 +1,14 @@
+import { useMemo } from 'react'
+import { useEthers as useDAppEthers } from "@usedapp/core";
+import { ethers } from "ethers";
+
+const { Web3Provider } = ethers.providers
+
+const useEthers = () => {
+  const { library, ...api } = useDAppEthers()
+  const isWeb3 = useMemo(() => library instanceof Web3Provider, [library])
```
Comment (sirpy) [inline packages/sdk-v2/src/hooks/useEthers.tsx:9]: why is this required?

**Review by sirpy — CHANGES_REQUESTED** [review]

File: packages/good-design/src/core/buttons/ActionButton.tsx (hunk 6 above)
Comment (L03TJ3) [reply packages/good-design/src/core/buttons/ActionButton.tsx:14]: This has been discussed and is being worked on as part of #16 
and there is an example app https://github.com/johnsmith-gooddollar/GoodDesignNativeBase
where Alex has defined a general default flow for so that we have consistent usage of nativebase/styling
is still to be tested

Comment (L03TJ3) [reply packages/good-design/src/core/buttons/ActionButton.tsx:14]: so ClaimButton comments should be done in #16
features got mixed during fixing of the branches last week or prior
